util/fetch_deps.ts

```diff
@@ -0,0 +1,229 @@
+// execute this script with `npm run fetch-deps`
+import fs from 'fs';
+import path from 'path';
+import crypto from 'crypto';
+import { promisify } from 'util';
+import { exec as _exec } from 'child_process';
+
+import _ from 'lodash';
+import tar from 'tar-fs';
+import JSZip from 'jszip';
+import json5 from 'json5';
+
+
+type Meta = {
+  'url': string;
+  'dest': string;
+  'strip': number;
+  'hash': [string, string];
+}
+
+const projectRoot = path.resolve(); // path of git repo
+
+const exec = promisify(_exec);
+
+const isFile = (p: string): boolean => {
+  if (!fs.existsSync(p))
+    return false;
+  return fs.statSync(p).isFile();
+};
+
+const isDir = (p: string): boolean => {
+  if (!fs.existsSync(p))
+    return false;
+  return fs.statSync(p).isDirectory();
+};
+
+const pad = (x: number) => x.toString().padStart(3, '0');
+
+const endsWith = (s: string, suffix: Iterable<string>): boolean => {
+  for (const ext of suffix) {
+    if (s.endsWith(ext))
+      return true;
+  }
+  return false;
+};
+
+const sleep = (time: number): void => {
+  const stop = new Date().getTime();
+  while (new Date().getTime() < stop + time)
+    ;
+
+  return;
+};
+
+export const safeRmDir = (path: string): void => {
+  let tries = 30;
+  while (tries > 0) {
+    tries--;
+    try {
+      fs.rmdirSync(path, { recursive: true });
+      break;
+    } catch (e) {
+      if (tries <= 0)
+        throw e;
+      sleep(0.3);
+    }
+  }
+};
+
+export const main = async (updateHashes = false): Promise<void> => {
+  const depsPath = path.join(projectRoot, 'util/DEPS.json5');
+
+  const deps: Record<string, Meta> = json5.parse(fs.readFileSync(depsPath).toString());
+
+  let cache: Record<string, Meta> = {};
+  const cachePath = path.join(projectRoot, 'util/DEPS.cache');
+  const dlPath = path.join(projectRoot, 'util/.deps_dl');
+
+  if (isFile(cachePath))
+    cache = JSON.parse(fs.readFileSync(cachePath).toString()) as Record<string, Meta>;
+
+  const oldCache = new Set(Object.keys(cache));
+  const newCache = new Set(Object.keys(deps));
+
+  const missing = new Set([...newCache].filter((v) => !oldCache.has(v)));
+  const obsolete = new Set([...oldCache].filter((v) => !newCache.has(v)));
+  const outdated: Set<string> = new Set();
+
+  for (const [key, meta] of Object.entries(deps)) {
+    if (!isDir(path.join(projectRoot, meta['dest'])))
+      missing.add(key);
+    else if (
+      ('hash' in meta) &&
+      (key in oldCache) &&
+      _.isEqual(_.get(cache, [key, 'hash'], [null, null]), meta['hash'][1])
+    )
+      outdated.add(key);
```

Comment:
I think `_.isEqual(_.get(cache, [key, 'hash'], [null, null]), meta['hash'][1])` is incorrect.  `_.get(cache, [key, 'hash'], [null, null])` is a `[algorithm, hash]` array and `meta['hash'][1]` is just the `hash`.  Additionally, I think you're looking for not equal?

Can I suggest for clarity:

```suggestion
    if (!isDir(path.join(projectRoot, meta['dest'])))
      missing.add(key);
    else if (oldCache.has(key) && cache[key]?.['hash'][1] !== meta['hash'][1])
      outdated.add(key);
```

I think this would also let you drop lodash as a dependency in this file, as `_has` and `_unset` seem pretty straightforward to write without lodash.

I think I'm not super keen on adding additional (logical) dependencies.  I agree that it is already an implicit npm dependency, but that doesn't mean that it is used in cactbot either other than one you just added.  It's still an additional knowledge burden.